Product SKU: Adds support for color, typography, and spacing

imanish003 opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This Pull Request enhances the styling capabilities of the Product SKU block by introducing additional support for color, typography & spacing. This improvement provides users with more flexibility when customizing the appearance of the Product SKU block.

Here is what Product SKU block supports now:

**Color:**
- Text color
- Background color

**Typography:**
- Font family 
- Font size
- Font Appearance
- Line height
- Letter spacing 
- Text decoration 
- Letter Case

**Spacing:**
- Margin
- Padding

<!-- Reference any related issues or PRs here -->

Part of #8691 & #7954

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### User Facing Testing

#### Test using blocks sidebar
- Create a new page or post
- Add the "Products" block to your post.
- Add Product SKU block as an inner block if not included already
- Select the Product SKU block and open the block settings in the right sidebar.
- Customize the block using the new styling options available:
   - Change the text color, background color, font family, font size, font appearance, line height, letter spacing, text decoration, letter case, margin, and padding under the appropriate settings sections.
- Verify that the styling changes are immediately visible and applied correctly within the editor itself.
- Save the post, then view it on the front-end to ensure the styling changes are visible and applied correctly.

#### Test using Global Styles
- Open the Global Styles sidebar by clicking the "Styles" button at the top right corner of the editor.

https://user-images.githubusercontent.com/16707866/228781997-24a9a29f-4be7-46d4-b1b3-41520692d86d.mov

- In the Global Styles sidebar, locate the "Product SKU" block.
- Customize the block using the new styling options available under the "Product SKU" block in the Global Styles sidebar:
   - Change the text color, background color, font family, font size, font appearance, line height, letter spacing, text decoration, letter case, margin, and padding under the appropriate settings sections.
- Verify that these applied correctly within editor & on frontend:
   - Create a new post
   - Add Products block
   - Add Product SKU block as an inner block if not included already
   - Check the Product SKU block in both the editor and on the front-end to ensure that the changes made through the Global styles UI are accurately applied:
       - In the editor, inspect Product SKU block & ensure it matches the values set in the Global styles UI.
       - Save the post & publish it. On frontend, inspect Product SKU block & ensure it matches the values set in the Global styles UI.

> Also, verify the changes in All Products block, which also uses Product SKU block.


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [X] Feature plugin
* [ ] Experimental

### Changelog

> Enhance Product SKU block styling capabilities with additional support for color, typography & spacing

imanish003 committed:
Add color, typography, and spacing support to Product SKU block

This commit introduces a new file `supports.ts` for the Product SKU block in the WooCommerce Blocks plugin, which adds color, typography, and spacing support. The following configuration has been added:

1. In `assets/js/atomic/blocks/product-elements/sku/supports.ts`, a new `supports` object is created, which extends the shared configuration and adds support for the following properties:
  - `text` and `background` color
  - `fontSize`, `lineHeight`, `__experimentalFontWeight`, `__experimentalFontFamily`, `__experimentalFontStyle`, `__experimentalTextTransform`, `__experimentalTextDecoration`, and `__experimentalLetterSpacing` for typography
  - `margin` and `padding` for spacing

imanish003 committed:
Add color, typography, and spacing support for Product SKU block in All Products block

This commit extends the support for color, typography, and spacing to the Product SKU block when used within the All Products block. The following changes have been made:

1. In `assets/js/atomic/blocks/product-elements/sku/block.tsx`, the `useColorProps`, `useSpacingProps`, and `useTypographyProps` hooks are imported and used to add the appropriate className and style properties to the rendered SKU element.
2. In `assets/js/atomic/blocks/product-elements/sku/edit.tsx`, the `style` property is separated from `useBlockProps()` and used directly on the `div` element for the SKU block.

imanish003 committed:
Merge branch 'trunk' into add/7954-product-sku-global-styles-review

imanish003 committed:
Add server-side rendering class for Product SKU Gutenberg block

This commit adds the ".wp-block-woocommerce-product-sku" class to the Product SKU Gutenberg block for server-side rendering purposes. The inclusion of this class ensures consistent styling between the editor and the front-end view of the block, providing a seamless experience for users.

The class name is generated based on the Gutenberg block prefix "wp-block", the namespace "woocommerce", and the block name "product-sku", creating a unique and identifiable class for the block. This class enables developers to target the block specifically in both the editor and front-end styles, maintaining the integrity of the block's appearance and functionality across different environments.

By adding this class to the server-side rendered version of the block, we ensure a unified and coherent styling experience throughout the block's usage.

imanish003 committed:
Fix minor issues

imanish003 commented:
@Aljullu, I am also adding you as reviewer because this PR adds `isDescendantOfAllProducts` flag for all products elements.
I noticed that product elements require special handling when used inside All products block. IMO this flag will be very useful in adding any new functionalities to product elements while still supporting All Products block.

I also considered creating a recursive function that checks block parents and their ancestors. Something like this:
```
function isDescendantOfAllProductsBlock(blockClientId, getBlockParents) {
  const parentClientIds = getBlockParents(blockClientId);

  if (parentClientIds.length === 0) {
    return false;
  }

  for (const parentId of parentClientIds) {
    const parentBlock = select('core/block-editor').getBlock(parentId);
    if (parentBlock.name === 'woocommerce/all-products') {
      return true;
    }

    if (isDescendantOfAllProductsBlock(parentId, getBlockParents)) {
      return true;
    }
  }

  return false;
}
```

But I dropped this idea because I thought this would have performance impact. Please let me know your thoughts about it 🙂

Aljullu commented on assets/js/atomic/blocks/product-elements/sku/supports.ts:
```diff
@@ -0,0 +1,37 @@
+ * External dependencies
+ */
+import { isFeaturePluginBuild } from '@woocommerce/block-settings';
+import { __experimentalGetSpacingClassesAndStyles } from '@wordpress/block-editor';
```
Heads-up that this line introduced an ESLint warning.

imanish003 replied:
ESLint warning is present wherever we use __experimentalGetSpacingClassesAndStyles, therefore I thought it was acceptable. However, I will add the following lines to ignore the warning going forward 🙂
```ts
// @ts-expect-error We check if this exists before using it.
// eslint-disable-next-line @wordpress/no-unsafe-wp-apis
```

kmanijak commented on assets/js/atomic/blocks/product-elements/sku/block.tsx:
```diff
@@ -9,6 +9,11 @@ import {
 } from '@woocommerce/shared-context';
 import { withProductDataContext } from '@woocommerce/shared-hocs';
 import type { HTMLAttributes } from 'react';
+import {
+	useColorProps,
+	useSpacingProps,
+	useTypographyProps,
+} from '@woocommerce/base-hooks';
```
I'm wondering if there's any other way of achieving that or can the hook be used on the Editor side?

The reason I'm asking is these hooks have a dependency on `@wordpress/block-editor` which has a dependency on `@wordpress/components` ([issue](https://github.com/woocommerce/woocommerce-blocks/issues/8071)) and neither of them should be used on the frontend because they're not designed for this and have a performance impact.

To be fair - it's not a blocker, because those hooks are already widely used on a frontend and given the `@wordpress` dependencies are externalized, there's no additional impact by yet another use of hooks (no matter the number of usages, the dependency is loaded just once). Having said that, I'm just raising this for awareness and maybe there's an easy way of avoiding adding another use of that (don't know that 😅 ).

imanish003 replied:
TBH, I was thinking about writing a P2 post about it. Ideally, we shouldn't need useColorProps, useSpacingProps, useTypographyProps hooks. 

Instead [here](https://github.com/woocommerce/woocommerce-blocks/blob/9e10039b4471566084ed3b99b808db92e3b713eb/assets/js/atomic/blocks/product-elements/save.js#L15) in save.js, we should return following instead:
```
export const Save = ( { attributes }: Props ): JSX.Element => {
	return (
		<div
			{ ...useBlockProps.save( {
				className: classnames( 'is-loading', attributes.className ),
			} ) }
		/>
	);
};
```

AFAIK `useBlockProps` hooks will generate `className` & `styles`, which we generate using these hooks(useColorProps, useSpacingProps, useTypographyProps). 
But I believe as we need to support product elements in `All products` block, we are using these hooks. 

@Aljullu Is my understanding correct? 🤔

Aljullu replied:
Just to clarify, these hooks are only loaded in the frontend in the _All Products_ block, correct? In that case, I would go ahead with this. If I'm not wrong, we have some other product element blocks that depend on those hooks as well. And sooner or later we will replace the _All Products_ block with the _Product_ block.

> I was thinking about writing a P2 post about it.

That would be super helpful. :bow:

kmanijak replied:
> If I'm not wrong, we have some other product element blocks that depend on those hooks as well.

That's true and as I wrote earlier "...I'm just raising this for awareness...". So I'm resolving the thread 👍

kmanijak commented on assets/js/atomic/blocks/product-elements/sku/edit.tsx:
```diff
@@ -18,7 +18,10 @@ const Edit = ( {
-	const blockProps = useBlockProps();
+	const { style, ...blockProps } = useBlockProps( {
+		className:
+			'wc-block-components-product-sku wp-block-woocommerce-product-sku',
```
Seems `wp-block-woocommerce-product-sku` is duplicated in a [block here](https://github.com/woocommerce/woocommerce-blocks/pull/8913/files#diff-60f7e4bda641c87212e5a65b413e1c794870354e6431235268e5794f784c893fR80). Can one of occurrances be removed?

imanish003 replied:
I apologize, but I am having trouble understanding your comment. Are you suggesting that `wc-block-components-product-sku` is equivalent to `wp-block-woocommerce-product-sku`? 🤔

kmanijak replied:
> Are you suggesting that wc-block-components-product-sku is equivalent to wp-block-woocommerce-product-sku? 🤔

Sorry I wasn't clear enough. What I mean is:

1. Here you provide a class name `wp-block-woocommerce-product-sku`, which is assigned to the `div` block container: [here](https://github.com/woocommerce/woocommerce-blocks/pull/8913/files#diff-273af84aa585a8e5a5a1446b8e97d85242eca59f5bf0696f3111b6627794ebaeR40)
2. The same class `wp-block-woocommerce-product-sku` is assigned to the Preview component [in here](https://github.com/woocommerce/woocommerce-blocks/pull/8913/files#diff-60f7e4bda641c87212e5a65b413e1c794870354e6431235268e5794f784c893fR80).

But I think I see what's happening, these are mutually exclusive, right? 🤔

imanish003 replied:
Thanks for the explanation. Now I can understand what you mean.

Actually, [here in block.tsx](https://github.com/woocommerce/woocommerce-blocks/blob/9e10039b4471566084ed3b99b808db92e3b713eb/assets/js/atomic/blocks/product-elements/sku/block.tsx#L77-L86), we are adding a class name for `All products`. Because classname added in [edit.tsx](https://github.com/woocommerce/woocommerce-blocks/blob/9e10039b4471566084ed3b99b808db92e3b713eb/assets/js/atomic/blocks/product-elements/sku/edit.tsx#L23) won't be added to `All products` blocks.

In summary, the class name in block.tsx will apply to the All products block, while the class name in edit.tsx will apply everywhere else. I hope that makes sense. 

Although I think this kind of handling shouldn't be required, but to support product elements in `All Products`, I added this 🙂

Aljullu commented on assets/js/atomic/blocks/product-elements/sku/supports.ts:
```diff
@@ -0,0 +1,37 @@
+			__experimentalTextDecoration: true,
+			__experimentalLetterSpacing: true,
+		},
+		__experimentalSelector: '.wc-block-components-product-sku',
```
Is `__experimentalSelector` required? I tried removing it and everything seemed to keep working as expected. :thinking:

imanish003 replied:
Good catch! I think we can remove `__experimentalSelector` flag.

Aljullu commented on assets/js/atomic/blocks/product-elements/sku/supports.ts:
```diff
@@ -0,0 +1,37 @@
+export const supports = {
+	...sharedConfig.supports,
+	...( isFeaturePluginBuild() && {
+		color: {
+			text: true,
+			background: true,
+		},
+		typography: {
+			fontSize: true,
+			lineHeight: true,
```
If we can remove `__experimentalSelector`, then we could make these styles available in WC core as well, so only the experimental ones would be gated to the feature plugin. :slightly_smiling_face:

imanish003 replied:
That makes sense. I will make the necessary changes 🤝

Aljullu reviewed (commented):
It's testing perfectly and code looks good to me. Nice improvement to the Product SKU block. :clap: 

Not sure if it's the expected behavior, but font-weight is only applied to the `SKU:` prefix, but not to the value:

![imatge](https://user-images.githubusercontent.com/3616980/229525567-6e6ce1b0-1b24-4a32-abd0-9d8c6351b34a.png)

> @Aljullu, I am also adding you as reviewer because this PR adds `isDescendantOfAllProducts` flag for all products elements.

I see we have similar flag for other parent blocks, so it makes sense to me. :slightly_smiling_face:

kmanijak reviewed (commented):
Thanks for taking care of this! Things I spotted while testing:


1. Little caveat (edge case I think) is the text is wrapped in the Editor while it may escape the container in the frontend. Below is the example in All Products, but it behaves the same in Products block.

| Editor | Frontend |
|---|---|
| <img width="414" alt="image" src="https://user-images.githubusercontent.com/20098064/229526370-598b0ca1-8eff-4457-b558-dbf7b4564464.png">| <img width="332" alt="image" src="https://user-images.githubusercontent.com/20098064/229526450-9253f01e-edb2-447f-ac24-4b9208dc5670.png"> |

2. Letter case is not reflected in the Editor but it works ok in Frontend while used in GLobal Styles. Below you can see it's set to lowercase in Editor, but it's kept uppercase in there.

| Editor | Frontend |
|---|---|
| <img width="652" alt="image" src="https://user-images.githubusercontent.com/20098064/229527969-7d4922f9-82a5-431a-9e74-e7ebafe3ee12.png">| <img width="386" alt="image" src="https://user-images.githubusercontent.com/20098064/229528017-ebeaa61b-1758-494c-a6f9-6f0011288609.png">|

Everything else looks good! I left some minor comments otherwise.

Also, the above may require the changes in Gutenberg, and in such case, I don't think they should block this PR from being merged 🙌

imanish003 committed:
Gate only experimental features behind isFeaturePluginBuild flag

This commit refactors the Product SKU block supports to enable color, typography, and spacing features for all builds, while keeping experimental features specific to the feature plugin build. It also adds necessary lint and TypeScript error handling for the experimental spacing API.

imanish003 commented:
@Aljullu 
> Not sure if it's the expected behavior, but font-weight is only applied to the SKU: prefix, but not to the value:

Actually, we have value text in a strong tag:
```
<strong>{ value }</strong>
```

I thought we wanna keep `strong` tag. I don't have a strong opinion about it, if you want, I can remove the strong tag. 🙂

> I see we have similar flag for other parent blocks, so it makes sense to me. 

Excellent, this will simplify the process of keeping product elements compatible with the All products block! 🙌🏻

imanish003 committed:
Add overflow-wrap to Product SKU block and update comment

This commit adds the overflow-wrap property to the Product SKU block styles to handle long SKU values gracefully.
It also updates the comment for the experimental spacing API to better reflect the usage check.

imanish003 committed:
Refactor Product SKU block classname and style handling

This commit refactors the Product SKU block classname and style handling. It combines classnames for better readability and removes the hardcoded textTransform style in favor of default styles.
This also fix the issue where textTransform style was not working as expected.

imanish003 committed:
Merge branch 'trunk' into add/7954-product-sku-global-styles-review

imanish003 committed:
Merge branch 'trunk' into add/7954-product-sku-global-styles-review

imanish003 committed:
Merge branch 'trunk' into add/7954-product-sku-global-styles-review

Aljullu reviewed (approved):
Looks good on my end, so approving. Leaving @kmanijak to do the final review on the items he brought up to the discussion.

> I thought we wanna keep `strong` tag. I don't have a strong opinion about it, if you want, I can remove the strong tag. 🙂

I don't have a specific preference either. I would say let's keep it as-is now. At some point it might be good to get design feedback on it, though.

imanish003 commented:
@Aljullu I appreciate your valuable input 🙏  I agree with you; we can keep the strong tag for now, and as you suggested, we can seek design feedback at a later stage. I'll wait for @kmanijak final review and address any concerns he may have before merging this PR 🙌🏻

kmanijak reviewed (approved):
> I'll wait for @kmanijak final review and address any concerns he may have before merging this PR 🙌🏻

Excellent work! I double checked the previously raised glitches and smoke tested - all works as expected. Great work @imanish003! 💪

imanish003 committed:
Merge branch 'trunk' into add/7954-product-sku-global-styles-review

opr commented:
Hey @imanish003 when editing the case of the SKU block, it also affects the SKU itself. Is this intended? Would we prefer to display the SKU exactly as the merchant entered in the product page?

imanish003 commented:
> Is this intended? Would we prefer to display the SKU exactly as the merchant entered in the product page?

This is intended as per the current implementation. I understand your concern about displaying the SKU as the merchant entered it. 
However, I'm not entirely sure whether we should display the SKU exactly as the merchant entered it because it is still possible to customize it the way merchant wants it using additional supports which were added in this PR. 

Perhaps we could get input from @shaunandrews on this matter? 🙂
